src/libiotc/iotc_roughtime_client.cc

```diff
@@ -0,0 +1,225 @@
+#define WOLFSSL_LEANPSK // 20190124
+// #ifndef __linux
+// #define __linux // 20190124
+// #endif
+
+#include <iotc_bsp_io_net.h>
+
+#include <arpa/inet.h>
+#include <assert.h>
+#include <errno.h>
+#include <inttypes.h>
+#include <netdb.h>
+#include <string.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include "client.h"
+#include "clock.h"
+#include "protocol.h"
+#include <iotc_bsp_rng.h>
+#include <iotc_bsp_time.h>
+#include <iotc_debug.h>
+#include <iotc_roughtime_client.h>
+#include <wolfssl/options.h>
+#include <wolfssl/ssl.h>
+#include <wolfssl/wolfcrypt/settings.h>
+
+// kTimeoutSeconds is the number of seconds that we will wait for a reply
+// from the server.
+static const int kTimeoutSeconds = 3;
+// kRetrialTimes is the number of time retrying to connect with server when if
+// it got failed
+static const int kRetrialTimes = 3;
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+namespace roughtime {
+
+iotc_roughtime_state_t
+iotc_roughtime_create_socket(int* out_socket, const char* server_address) {
+  std::string address(server_address);
+
+  const size_t kColonOffset = address.rfind(':');
+  if (kColonOffset == std::string ::npos) {
+    iotc_debug_format(stderr, "No port number in server address: %s\n",
+                      address.c_str());
+    return IOTC_ROUGHTIME_ERROR;
+  }
+  std::string host(address.substr(0, kColonOffset));
+  const std::string kPortStr(address.substr(kColonOffset + 1));
+  int port = std::stoi(kPortStr.c_str());
+
+  if (IOTC_BSP_IO_NET_STATE_OK !=
+      iotc_bsp_io_net_socket_connect(
+          reinterpret_cast<iotc_bsp_socket_t*>(out_socket), host.c_str(), port,
+          SOCKET_DGRAM)) {
+    iotc_debug_logger("Connect to the socket");
+  }
+
+  if (IOTC_BSP_IO_NET_STATE_OK !=
+      iotc_bsp_io_net_connection_check(*out_socket, host.c_str(), port)) {
+    iotc_debug_logger("Check the connection");
+  }
+
+  return IOTC_ROUGHTIME_OK;
+}
+
+iotc_roughtime_state_t
+iotc_roughtime_receive_time(int socket, const char* name,
+                            const char* public_key,
+                            iotc_roughtime_timedata_t* time_data) {
+
+  std::string server_name(name);
+  std::string server_public_key(public_key);
+  uint8_t nonce[roughtime::kNonceLength];
+  size_t socket_evts_size = 1;
+  iotc_bsp_io_net_state_t state;
+  iotc_bsp_socket_events_t socket_evts[socket_evts_size];
```

Comment:
I'd simplify by using a simple variable instead of an array of one:
`iotc_bsp_socket_events_t socket_event;`